ui/src/vertical-geometry/km-heights-fetch.ts

```diff
@@ -0,0 +1,242 @@
+}
+
+function getCache(changeTimes: ChangeTimes, alignmentId: VerticalGeometryDiagramAlignmentId) {
+    return 'planId' in alignmentId
+        ? geometryHeightsCache(toDate(changeTimes.geometryPlan))
+        : locationTrackHeightsCache(toDate(changeTimes.layoutLocationTrack));
+}
+
+function getCacheItem(
+    changeTimes: ChangeTimes,
+    alignmentId: VerticalGeometryDiagramAlignmentId,
+    tickLength: number,
+): HeightCacheItem | undefined {
+    const cache = getCache(changeTimes, alignmentId);
+    const key = heightCacheKey(alignmentId, tickLength);
+    return cache.get(key);
+}
+
+function getOrCreateCacheItem(
+    changeTimes: ChangeTimes,
+    alignmentId: VerticalGeometryDiagramAlignmentId,
+    tickLength: number,
+): HeightCacheItem {
+    const cache = getCache(changeTimes, alignmentId);
+    const key = heightCacheKey(alignmentId, tickLength);
+    const item = cache.get(key);
+    if (item !== undefined) {
+        return item;
+    } else {
+        const newItem = {
+            resolved: [],
+        };
+        cache.set(key, newItem);
+        return newItem;
+    }
+}
+
+export function weaveKms<T>(getFirstM: (km: T) => number, left: T[], right: T[]): T[] {
+    const rv: T[] = [];
+    let leftI = 0,
+        rightI = 0;
+
+    while (leftI < left.length && rightI < right.length) {
+        const leftM = getFirstM(left[leftI]);
+        const rightM = getFirstM(right[rightI]);
+        if (leftM < rightM) {
+            rv.push(left[leftI++]);
+        } else if (rightM < leftM) {
+            rv.push(right[rightI++]);
+        } else {
+            rv.push(left[leftI++]);
+            rightI++;
+        }
+    }
+    if (leftI >= left.length) {
+        rv.push(...right.slice(rightI));
+    } else if (rightI >= right.length) {
+        rv.push(...left.slice(leftI));
+    }
+    return rv;
+}
+
+async function loadAlignmentHeights(
+    alignmentId: VerticalGeometryDiagramAlignmentId,
+    startM: number,
+    endM: number,
+    tickLength: number,
+): Promise<[TrackKmHeights[], [VerticalGeometryDiagramAlignmentId, number]]> {
+    return (
+        'planId' in alignmentId
+            ? getPlanAlignmentHeights(
+                  alignmentId.planId,
+                  alignmentId.alignmentId,
+                  startM,
+                  endM,
+                  tickLength,
+              )
+            : getLocationTrackHeights(
+                  alignmentId.locationTrackId,
+                  alignmentId.publishType,
+                  startM,
+                  endM,
+                  tickLength,
+              )
+    ).then((r) => [r, [alignmentId, tickLength]]);
+}
+
+export function getMissingCoveringRange(
+    ranges: [number, number][],
+    queryStart: number,
+    queryEnd: number,
+): null | [number, number] {
+    const leftStart = ranges.findIndex(([_, rangeEnd]) => rangeEnd >= queryStart);
+    if (leftStart !== -1) {
+        for (let i = 0; i < ranges.length; i++) {
```

Comment:
Tässä ei näyttäisi olevan tarvetta manuaaliselle indeksille, eli voisi olla vain `for (range in ranges) {...}`. Vai olikos tämän tarkoitus aloittaa leftStartista eikä 0:sta?